akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/RollingUpdateShardAllocationSpec.scala

```diff
@@ -0,0 +1,193 @@
+
+  "Cluster sharding" must {
+
+    "form cluster" in {
+      awaitClusterUp(first, second)
+      enterBarrier("cluster-started")
+    }
+    lazy val shardRegion = startSharding(
+      system,
+      typeName = typeName,
+      entityProps = Props[GiveMeYourHome](),
+      extractEntityId = GiveMeYourHome.extractEntityId,
+      extractShardId = GiveMeYourHome.extractShardId)
+
+    "start cluster sharding on first" in {
+      runOn(first, second) {
+        shardRegion
+        shardRegion ! GiveMeYourHome.Get("id1")
+        // started on either of the nodes
+        val address1 = expectMsgType[GiveMeYourHome.Home].address
+
+        shardRegion ! GiveMeYourHome.Get("id2")
+        // started on the other of the nodes (because least
+        val address2 = expectMsgType[GiveMeYourHome.Home].address
+
+        // one on each node
+        Set(address1, address2) should have size (2)
+      }
+      enterBarrier("first-version-started")
+    }
+    "start a rolling upgrade" in {
+      join(third, first)
+
+      runOn(first, second, third) {
+        shardRegion
+
+        // new shards should now go on third since that is the highest version,
+        // however there is a race where the shard has not yet completed registration
+        // with the coordinator and shards will be allocated on the old nodes, so we need
+        // to make sure the third region has completed registration before trying
+        // if we didn't the strategy will default it back to the old nodes
+        awaitAssert {
+          shardRegion ! ShardRegion.GetCurrentRegions
+          expectMsgType[ShardRegion.CurrentRegions].regions should have size (3)
+        }
+      }
+      enterBarrier("third-region-registered")
+      runOn(first, second) {
+        shardRegion ! GiveMeYourHome.Get("id3")
+        expectMsgType[GiveMeYourHome.Home]
+      }
+      runOn(third) {
+        // now third region should be only option as the other two are old versions
+        shardRegion ! GiveMeYourHome.Get("id3")
+        expectMsgType[GiveMeYourHome.Home].address should ===(Cluster(system).selfAddress)
+      }
+      enterBarrier("rolling-upgrade-in-progress")
+    }
+    "complete a rolling upgrade" in {
+      join(fourth, first)
+
+      runOn(first) {
+        val cluster = Cluster(system)
+        cluster.leave(cluster.selfAddress)
+      }
+      enterBarrier("first-left")
+
+      runOn(second, third, fourth) {
+        awaitAssert {
+          shardRegion ! ShardRegion.GetCurrentRegions
+          expectMsgType[ShardRegion.CurrentRegions].regions should have size (3)
+        }
+      }
+      enterBarrier("sharding-handed-off")
+
+      // trigger allocation (we don't know which id was on node 1)
+      runOn(second, third, fourth) {
+        awaitAssert {
+          shardRegion ! GiveMeYourHome.Get("id1")
+          expectMsgType[GiveMeYourHome.Home]
+
+          shardRegion ! GiveMeYourHome.Get("id2")
+          expectMsgType[GiveMeYourHome.Home]
```

Comment:
This second phase is not asserting much. Maybe check that id1 and id2 end up at third or forth in the end